straxen/plugins/peak_processing.py

```diff
@@ -331,98 +336,393 @@ def find_n_competing(peaks, windows, fraction):
+        return 10 * self.config['ambience_time_window_backward']
+
+    @property
+    def origin_dtype(self):
+        return ['lh_before', 's0_before', 's1_before', 's2_before', 's2_near']
+
+    def infer_dtype(self):
+        dtype = []
+        for ambience in self.origin_dtype:
+            dtype += [((f"Number of small {' '.join(ambience.split('_'))} a peak", 
+                        f'n_{ambience}'), np.int16), 
+                      ((f"Area sum of small {' '.join(ambience.split('_'))} a peak", 
+                        f's_{ambience}'), np.float32)]
+        dtype += strax.time_fields
+        return dtype
+
+    def compute(self, lone_hits, peaks):
+        return self.compute_ambience(lone_hits, peaks, peaks)
+
+    def compute_ambience(self, lone_hits, peaks, current_peak):
+        # 1. Initialization
+        result = np.zeros(len(current_peak), self.dtype)
+        num_array = np.zeros(len(current_peak), dtype=np.int16)
+        sum_array = np.zeros(len(current_peak), dtype=np.float32)
+
+        # 2. Define time window for each peak, we will find small peaks & lone hits within these time windows
+        roi = np.zeros(len(current_peak), dtype=strax.time_fields)
+        roi['time'] = current_peak['center_time'] - self.config['ambience_time_window_backward']
+        roi['endtime'] = current_peak['center_time']
+
+        # 3. Calculate number and area sum of lonehits before a peak
+        touching_windows = strax.touching_windows(lone_hits, roi)
+        # Calculating ambience
+        self.lonehits_ambience(current_peak, 
+                               lone_hits, 
+                               touching_windows, 
+                               num_array, 
+                               sum_array, 
+                               self.config['ambience_divide_t'])
+        result['n_lh_before'] = num_array
+        result['s_lh_before'] = sum_array
+
+        # 4. Calculate number and area sum of small S0, S1, S2 before a peak
+        radius = -1
+        for stype, area in zip([0, 1, 2], 
+                                         self.config['ambience_area_parameters']):
+            num_array = np.zeros(len(current_peak), dtype=np.int16)
+            sum_array = np.zeros(len(current_peak), dtype=np.float32)
+            mask_pre = (peaks['type'] == stype) & (peaks['area'] < area)
+            touching_windows = strax.touching_windows(peaks[mask_pre], roi)
+            # Calculating ambience
+            self.peaks_ambience(current_peak, 
+                                peaks[mask_pre], 
+                                touching_windows, 
+                                radius, 
+                                num_array, 
+                                sum_array, 
+                                self.config['ambience_divide_t'], 
+                                self.config['ambience_divide_r'])
+            result[f'n_s{stype}_before'] = num_array
+            result[f's_s{stype}_before'] = sum_array
```

Comment:
I think for a future version it would be better to absorb your stype into your `peak_ambience` function. Such that in your peak ambience function you have three buffers for S0, S1, S2 and you are adding the values depending on the type. E.g. sum_array would look like:

```
sum_array = np.zeros((len(peaks), 3))
# adding sum result:
sum_array[p_i, creating_peak['type']] = your_result

```

 In this way you only have to loop over all peaks once instead of 3 times.

Reply:
We would have to pay a bit attention to your different thresholds though.